deepchem/models/torch_models/layers.py

```diff
@@ -4187,3 +4189,207 @@ def gaussian_histogram(self, x: torch.Tensor) -> torch.Tensor:
+
+        self.x_edge_mlp: MultilayerPerceptron = MultilayerPerceptron(
+            d_input=self.dim * 3, d_output=self.dim, activation_fn='silu')
+        self.linear: nn.Linear = nn.Linear(self.dim, self.dim, bias=False)
+
+    def forward(self, node_features: torch.Tensor,
+                edge_attributes: torch.Tensor,
+                edge_indices: torch.Tensor) -> torch.Tensor:
+        """
+        Performs the forward pass of the GlobalMessagePassing layer.
+
+        Parameters
+        -----------
+        node_features: torch.Tensor
+            The input node features tensor of shape (num_nodes, feature_dim).
+        edge_attributes: torch.Tensor
+            The input edge attribute tensor of shape (num_edges, attribute_dim).
+        edge_indices: torch.Tensor
+            The input edge index tensor of shape (2, num_edges).
+
+        Returns
+        --------
+        torch.Tensor
+            The updated node features tensor after message passing of shape (num_nodes, feature_dim).
+        """
+        edge_indices, _ = add_self_loops(edge_indices,
+                                         num_nodes=node_features.size(0))
+
+        residual_node_features: torch.Tensor = node_features
+
+        # Integrate the Cross Layer Mapping inside the Global Message Passing
+        node_features = self.h_mlp(node_features)
+
+        # Message Passing operation
+        node_features = self.propagate(edge_indices,
+                                       x=node_features,
+                                       num_nodes=node_features.size(0),
+                                       edge_attr=edge_attributes)
+
+        # Update function f_u
+        node_features = self.res1(node_features)
+        node_features = self.mlp(node_features) + residual_node_features
+        node_features = self.res2(node_features)
+        node_features = self.res3(node_features)
+
+        # Message Passing operation
+        node_features = self.propagate(edge_indices,
+                                       x=node_features,
+                                       num_nodes=node_features.size(0),
+                                       edge_attr=edge_attributes)
+
+        return node_features
+
+    def message(self, x_i: torch.Tensor, x_j: torch.Tensor,
+                edge_attr: torch.Tensor) -> torch.Tensor:
+        """Constructs messages to be passed along the edges in the graph.
+
+        Parameters
+        -----------
+        x_i: torch.Tensor
+            The source node features tensor of shape (num_edges, feature_dim).
+        x_j: torch.Tensor
+            The target node features tensor of shape (num_edges, feature_dim).
+        edge_attributes: torch.Tensor
+            The edge attribute tensor of shape (num_edges, attribute_dim).
+
+        Returns
+        --------
+        torch.Tensor
+            The constructed messages tensor.
+        """
+        num_edge: int = edge_attr.size()[0]
```

Comment:
Is there a case where `x_i` will have a different size at dimension 0 compared to `edge_attr`? I am trying to understand the thought process behind `num_edge`

Reply:
Actually there was a mistake, the shape of `x_i` and `x_j` is (num_edges+num_nodes, feature_dim) due to the addition of self_loops. Variable `num_edge` is equal to num_edges and `x_i[:num_edge],` `x_j[:num_edge],` and `edge_attr` is being concatenated. While the self_loops are being handled separately.

Reply:
Ah sure, that makes sense!